analyses/cms-open-data-ttbar/validate_histograms.py

```diff
@@ -42,8 +43,57 @@ def validate(histos: dict, reference: dict) -> dict[str, list[str]]:
         if not np.allclose(h['edges'], ref_h['edges']):
             errors[name].append(f"Edges do not match:\n\tgot      {h['edges']}\n\texpected {ref_h['edges']}")
         contents_depend_on_rng = "pt_res_up" in name # skip checking the contents of these histograms as they are not stable
-        if not contents_depend_on_rng and not np.allclose(h['contents'], ref_h['contents']):
-            errors[name].append(f"Contents do not match:\n\tgot      {h['contents']}\n\texpected {ref_h['contents']}")
+        is_close = np.isclose(h['contents'], ref_h['contents']) 
+        
+        #### check if bin migration ####
+        if not contents_depend_on_rng and not all(is_close):
+
+            # gets indices where above array is false
+            where_not_close = np.where(np.invert(is_close))[0]
+            # gets difference of adjacent entries
+            diff_values = np.diff(where_not_close)
+            # get the indices where the above difference is greater than 1 (to form groupings of bins based on location)
+            split_indices = np.argwhere(np.abs(diff_values)>1)
+            # np.argwhere adds extra dimension we don't need
+            split_indices = split_indices.reshape((split_indices.shape[0],))
+            
+            # if only one grouping detected
+            if len(split_indices)==0:
+                split_values = [where_not_close]
+            # if more than one grouping detected
+            else:
+                # shift indices to the point we want to split at
+                split_indices = split_indices + 1
+                # split indices into groupings
+                split_values = np.split(where_not_close, split_indices)
+            
+            is_error=False
+            # iterate through groupings and test for bin migration or error
+            for group in split_values:
+                h_group = np.array(h['contents'])[group]
+                ref_group = np.array(ref_h['contents'])[group]
+                # if difference is great, count as error
+                if not np.allclose(h_group, ref_group, atol=2.0):
```

Comment:
`2.0` looks magical, maybe add a not about where it comes from (e.g. point to the GitHub issues where there are measurements on the impact of bin migrations)

Reply:
We briefly talked about this today as well, it is indeed a pretty magical number and if we have a reference that would be useful. It is most strict with very few files (assuming migrations are rare to begin with), as with more files each event will have a smaller weight and it would require the same proportion of additional events migrating to stick with the same absolute migration. I assume that people would mostly compare setups with few files, for which a value of 2 doesn't feel unreasonable.

Reply:
That's totally ok, I was just suggesting to add a comment to explain this (or point to an explanation) to someone reading the code, because they might need to know where the 2. came from to decide whether it still makes sense in 5 years or for a slightly different use case

Reply:
For sure, I agree with the comment.